design/src/rfcs/rfc0012_crates_io_publishing.md

```diff
@@ -0,0 +1,214 @@
+so all model changes will result in a `minor` version bump during this phase.
+
+Overall, determining a generated crate's version number looks as follows:
+
+![Phase 1: How to version a generated crate](rfc0012_crates_io_publishing/phase1_generated_crate_version.svg)
+
+- __A: smithy-rs changed?__: Compare the `smithy_rs_version` in the previous `versions.toml` with the
+  next `versions.toml` file, and if the values are different, consider [smithy-rs] to have changed.
+- __B: model changed?__: Similarly, compare the `model_hash` for the crate in `versions.toml`.
+
+### Versioning for Runtime Crates
+
+The old scheme of all runtime crates in [smithy-rs] having a fake `0.0.0-smithy-rs-head` version number with
+a build step to replace those with a consistent `major.minor` will be removed. These runtime crates will begin
+having their actual next version number in the Cargo.toml file in smithy-rs.
+
+This introduces a new problem where a developer can forget to bump a runtime crate version, so a method of
+process enforcement needs to be introduced. This will be done through CI when merging into `smithy-rs/main`
+and repeated when merging into `aws-sdk-rust/main`.
+
+The following checks need to be run for runtime crates:
+
+![Phase 1: How to validate a runtime version bump](rfc0012_crates_io_publishing/phase1_runtime_crate_version_checks.svg)
+
+- __A: Crate has changed?__ The crate's source files and manifest will be hashed for the previous version
+  and the next version. If these hashes match, then the crate is considered unchanged.
+- __B: Minor bumped?__ The previous version is compared against the next version to see if the minor version
+  number was bumped.
+- __C: Version changed?__ The previous version is compared against the next version to see if it changed.
+- __D: Patch bumped?__ The previous version is compared against the next version to see if the patch version
+  number was bumped.
+- __E: Semverver passes?__ Runs [rust-semverver] against the old and new versions of the crate.
+  - If semverver fails to run (for example, if it needs to be updated to the latest nightly to succeed),
+    then fail CI saying that either semverver needs to be updated, or that a minor version bump is required.
+  - If semverver results in errors, fail CI indicating a minor version bump is required.
+  - If semverver passes, then pass CI.
+
+For CI, the `aws-sdk-rust/main` branch's `versions.toml` file is the source of truth for the previous release's
+crate versions and source code.
+
+### Yanking
+
+The publisher tool will be updated to read the `versions.toml` to yank all versions published in a release.
+This process will look as follows:
+1. Take a path to a local clone of the [aws-sdk-rust] repository
+2. Confirm the branch is `main` and that HEAD is tagged as a release
+3. Read `versions.toml` and print out summary of crates to yank
+4. Confirm with user before proceeding
+5. Yank crates
```

Comment:
I'm a little confused about this—I think I'm missing how versions.toml tells you which crates were released and which were not given the current schema—do you have to diff the files?

Reply:
Let's say we have a bad release at tag `bad` that we want to yank. The `versions.toml` file at that tag will have all the crate version numbers associated with that release, but if the publish failed half way through, then some of those might be published to crates.io while others might not be. But that's OK since the yank tool can just check that the version is actually published using the crates.io API before yanking it.

That said, I'm now confused by (2) above. I think I'll revise this to make sure you're on a release tag rather than on `main`.